### DISCUSSION TITLE
Ability to change content collection's primary key property name

### DISCUSSION THREAD

**Discussion by Souvlaki42** [withastro/roadmap#1082] [Proposal]
### Body

# Summary

What about the ability to change how the primary key/id property is called in a content collection.

# Background & Motivation

This isn't something I absolutely need but it could be a nice improvement. I was loading a JSON file with data inside a content collection in Astro 5.1 the other day. This dataset didn't have a property called id but a unique property that could be used as an ID without being called that way. I needed the change that property key name to id despite having a more descriptive name.

# Goals

- Being able to use any property you want as an ID without calling it that.

# Example

```ts
const elements = defineCollection({
  loader: file("src/elements.json"),
  primaryKey: "atomicNumber",
  schema: z.object({ atomicNumber: z.number(), ... })
});
```

**Comment by florian-lefebvre** [comment]
A workaround is to pass a custom parse function, where you return an id key duplicated of another property

**Comment by Souvlaki42** [comment]
@florian-lefebvre Thanks for your response. This could work but there are are some issues with it. This is my content.config.ts now:
```ts
import { file } from "astro/loaders";
import { defineCollection, z } from "astro:content";
import { elementSchema } from "~/lib/schemas";

type Element = z.infer<typeof elementSchema>;
const elements = defineCollection({
  loader: file("src/elements.json", {
    parser: (text) => {
      const json = JSON.parse(text) as Element[];
      return json.map((element) => {
        return {
          ...element,
          id: element.atomicNumber
        };
      });
    }
  }),
  schema: elementSchema
});

export const collections = { elements };
```
Typescript didn't complain but still got the error:
```
23:37:54 [ERROR] [file-loader] Item in src/elements.json is missing an id or slug field.
```
Also JSON parse returns any and had to work around it. Lastly, this isn't documented anywhere for people that don't know. My solution just seems simpler. I am even positive of at least trying to implement it myself into the codebase and make a pull request for it. What do you think?

**Reply by florian-lefebvre** [reply]
For the error I think that may be because id must be a string. But I agree with you this is just a workaround. Many people are on holiday rn so do not work on a PR for now, probably wanna have @ascorbic opinion on this

**Reply by Souvlaki42** [reply]
Understandable. I'll wait then.
Merry Christmas!

**Reply by ascorbic** [reply]
Do you have `id` in the schema?

**Reply by Souvlaki42** [reply]
> Do you have `id` in the schema?

I had but it was a number and now it is a string:
```ts
// ./src/lib/schemas.ts
export const elementSchema = z.object({
  id: z.string(),
  atomicNumber: z.number(),
  symbol: z.string(),
  name: z.string(),
  atomicMass: z.string().or(z.array(z.number())),
  cpkHexColor: z.string().or(z.number()),
  electronicConfiguration: z.string(),
  electronegativity: z.string().or(z.number()),
  atomicRadius: z.string().or(z.number()),
  ionRadius: z.string(),
  vanDerWaalsRadius: z.string().or(z.number()),
  ionizationEnergy: z.string().or(z.number()),
  electronAffinity: z.string().or(z.number()),
  oxidationStates: z.string().or(z.number()),
  standardState: z.string(),
  bondingType: z.string(),
  meltingPoint: z.string().or(z.number()),
  boilingPoint: z.string().or(z.number()),
  density: z.string().or(z.number()),
  groupBlock: z.string(),
  yearDiscovered: z.string().or(z.number())
});

// ./src/content.config.ts
import { file } from "astro/loaders";
import { defineCollection, z } from "astro:content";
import { elementSchema } from "~/lib/schemas";

type Element = z.infer<typeof elementSchema>;
const elements = defineCollection({
  loader: file("src/elements.json", {
    parser: (text) => {
      const json = JSON.parse(text) as Element[];
      return json.map((element) => {
        return {
          ...element,
          id: element.atomicNumber.toString()
        };
      });
    }
  }),
  schema: elementSchema
});

export const collections = { elements };

// ./src/pages/[atomic].astro
export async function getStaticPaths() {
  const elements = await getCollection("elements");

  return elements.map((element) => ({
    params: { atomic: element.data.atomicNumber },
    props: { rawElement: element }
  }));
}

```
Now I am getting:
```
using deprecated parameters for the initialization function; pass a single object instead
19:20:08 [ERROR] [file-loader] Error reading data from src/elements.json
```
This is still a workaround btw. This is why I want this feature. What do you think?

**Reply by ascorbic** [reply]
I'm afraid I have no idea where that error is coming from. It's not an Astro error. If you could create a minimal repro I'll take a look.

**Reply by Souvlaki42** [reply]
I tried. I failed but in the process I learned that this `using deprecated parameters for the initialization function; pass a single object instead` is propably related to qwik dev integration with astro I used and this `19:20:08 [ERROR] [file-loader] Error reading data from src/elements.json` existed because of a change to the json file I did a long ago. It's fixed now but this shouldn't be something I would need to do myself.
